websearch/brave/src/lib.rs

```diff
@@ -0,0 +1,135 @@
+mod client;
+mod conversions;
+
+use std::cell::RefCell;
+
+use crate::client::{BraveSearchApi, SearchRequest};
+use crate::conversions::{params_to_request, response_to_results, validate_search_params};
+use golem_web_search::golem::web_search::web_search::{
+    Guest, GuestSearchSession, SearchError, SearchMetadata, SearchParams, SearchResult,
+    SearchSession,
+};
+
+use golem_web_search::LOGGING_STATE;
+
+struct BraveSearch {
+    client: BraveSearchApi,
+    request: SearchRequest,
+    params: SearchParams,
+    finished: bool,
+    metadata: Option<SearchMetadata>,
+}
+
+impl BraveSearch {
+    fn new(client: BraveSearchApi, request: SearchRequest, params: SearchParams) -> Self {
+        Self {
+            client,
+            request,
+            params,
+            finished: false,
+            metadata: None,
+        }
+    }
+
+    fn next_page(&mut self) -> Result<Vec<SearchResult>, SearchError> {
+        if self.finished {
+            return Ok(vec![]);
+        }
+
+        let response = self.client.search(self.request.clone())?;
+        let (results, metadata) = response_to_results(response, &self.params);
+
+        self.metadata = metadata;
+        self.finished = true;
```

Comment:
This will only return a single page from what I can tell.

You'll need to adjust offset by one to get the next page and only set finished if more_results_available in the response is false.